### PR TITLE
Network parameters

### PR DESCRIPTION
**Pull request by IncognitoJam**
### Motivation
Some constants for Bitcoin are currently defined in [`bitcoin/types.go`](https://github.com/rosetta-dogecoin/rosetta-dogecoin/blob/main/bitcoin/types.go#L95) but most of the network parameters are in [`btcd/chaincfg`](https://github.com/btcsuite/btcd/blob/master/chaincfg/params.go#L225). So, we need some place to define these for Dogecoin.

### Solution
I think the easiest way to do this is to just copy and paste the relevant definitions from `btcd` into our own package, which I've started in this PR:
- `genesis.go`: definitions for the genesis block, genesis tx etc.
- `params.go`: other network parameters like target time, checkpoints, deployments...

Then we can modify each of these definitions to match the dogecoin source.

### Open questions
I'm looking for feedback on this! There might be a better way to structure it...

### CODE REVIEW

**Comment by xanimo** [comment]
Sounds good to me!

**Comment by patricklodder** [comment]
Re: specifying it in the dogecoin module - I think this is the right way, if we can.

Question back: How would we pass the btcd parametrization to btcd? Is there / do you have a proposed integration path for that?

**Comment by IncognitoJam** [comment]
I think we can replace references to the bitcoin package with the dogecoin package (like replace `bitcoin.MainnetNetwork` with `dogecoin.MainnetNetwork`) for the params we want to override, I've just tried this out to see how much we would have to change (almost 100 LOC).

**Comment by patricklodder** [comment]
This looks like a workable solution to me.

**Comment by IncognitoJam** [comment]
The first three commits in this PR were just setup and copying the Bitcoin configuration, and the final two are modifying everything for Dogecoin. Would it help if this PR was split in two to make it easier to review?

I've made my best effort at getting all of the params right but there are a few things I couldn't find in the Dogecoin source

File: dogecoin/params.go [hunk 1]
```diff
@@ -141,9 +127,9 @@ var MainNetParams = chaincfg.Params{
...
 
 	// Address encoding magics
-	PubKeyHashAddrID:        0x00, // starts with 1
-	ScriptHashAddrID:        0x05, // starts with 3
-	PrivateKeyID:            0x80, // starts with 5 (uncompressed) or K (compressed)
+	PubKeyHashAddrID:        0x1E, // 30, starts with 1
+	ScriptHashAddrID:        0x16, // 22, starts with 3
+	PrivateKeyID:            0x9E, // 158, starts with 5 (uncompressed) or K (compressed)
 	WitnessPubKeyHashAddrID: 0x06, // starts with p2
 	WitnessScriptHashAddrID: 0x0A, // starts with 7Xh
```
Comment (IncognitoJam) [inline dogecoin/params.go:134]: Address encoding:
- Need to update comments, not sure what the correct values are other than D for the pubkey one?
- Not sure what the values should be for Witness

Comment (patricklodder) [reply dogecoin/params.go:134]: - starts with D
- starts with 9 or A
- starts with Q

File: dogecoin/params.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+		{Height: 2750000, Hash: newHashFromStr("473ea9f625d59f534ffcc9738ffc58f7b7b1e0e993078614f5484a9505885563")},
+		{Height: 3062910, Hash: newHashFromStr("113c41c00934f940a41f99d18b2ad9aefd183a4b7fe80527e1e6c12779bd0246")},
+	},
+
+	// Consensus rule change deployments.
+	//
+	// The miner confirmation window is defined as:
+	//   target proof of work timespan / target proof of work spacing
+	RuleChangeActivationThreshold: 9576,  // 95% of MinerConfirmationWindow
+	MinerConfirmationWindow:       10080, //
```
Comment (patricklodder) [inline dogecoin/params.go:103]: Those were never used in core and have until now always been 1900 and 2000 respectively.

File: dogecoin/params.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+			ExpireTime: 0,          // Disabled
+		},
+	},
+
+	// Mempool parameters
+	RelayNonStdTxs: false,
+
+	// Human-readable part for Bech32 encoded segwit addresses, as defined in
+	// BIP 173.
+	Bech32HRPSegwit: "bc", // always bc for main net
```
Comment (patricklodder) [inline dogecoin/params.go:127]: Although currently unimplemented, this is planned to be "doge" in 0.21, see [chainparams.cpp in 8b0dcb5 on dogecoin/dogecoin](https://github.com/dogecoin/dogecoin/blob/8b0dcb523dcda768988f0ed661a70611c3d727b3/src/chainparams.cpp#L137)

File: dogecoin/params.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	// Address encoding magics
+	PubKeyHashAddrID:        0x1E, // 30, starts with 1
+	ScriptHashAddrID:        0x16, // 22, starts with 3
+	PrivateKeyID:            0x9E, // 158, starts with 5 (uncompressed) or K (compressed)
+	WitnessPubKeyHashAddrID: 0x06, // starts with p2
+	WitnessScriptHashAddrID: 0x0A, // starts with 7Xh
+
+	// BIP32 hierarchical deterministic extended key magics
+	HDPrivateKeyID: [4]byte{0x04, 0x88, 0xad, 0xe4}, // starts with xprv
+	HDPublicKeyID:  [4]byte{0x04, 0x88, 0xb2, 0x1e}, // starts with xpub
```
Comment (patricklodder) [inline dogecoin/params.go:138]: comment: dgpv and dgub respectively

File: dogecoin/params.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+	// BIP44 coin type used in the hierarchical deterministic path for
+	// address generation.
+	HDCoinType: 0,
```
Comment (patricklodder) [inline dogecoin/params.go:142]: BIP44 coin is 3 for DOGE

**Review by patricklodder — COMMENTED** [review]
Found some minor things

**Comment by IncognitoJam** [comment]
@patricklodder I've made some changes to reflect your comments, thanks!

**Comment by patricklodder** [comment]
Alright. All looks good. I'll do some local testing after rebase and we should be good to go with this! Thank you! 👍

**Comment by IncognitoJam** [comment]
Done! salute

**Review by patricklodder — APPROVED** [review]

**Pull request merged**